File: openfeature-provider/go/confidence/grpc_wasm_flag_logger.go
package confidence

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	resolverv1 "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/proto/confidence/flags/resolverinternal"
	"google.golang.org/grpc/metadata"
)

const (
	// MaxFlagAssignedPerChunk is the max number of flag_assigned entries per chunk
	// to avoid exceeding gRPC max message size
	MaxFlagAssignedPerChunk = 1000
)

type FlagLogger interface {
	Write(ctx context.Context, request *resolverv1.WriteFlagLogsRequest) error
	Shutdown()
}

type GrpcFlagLogger struct {
	stub         resolverv1.InternalFlagLoggerServiceClient
	clientSecret string
	logger       *slog.Logger
	wg           sync.WaitGroup
}

// Compile-time interface conformance check
var _ FlagLogger = (*GrpcFlagLogger)(nil)

func NewGrpcWasmFlagLogger(stub resolverv1.InternalFlagLoggerServiceClient, clientSecret string, logger *slog.Logger) *GrpcFlagLogger {
	return &GrpcFlagLogger{
		stub:         stub,
		clientSecret: clientSecret,
		logger:       logger,
	}
}

// Write writes flag logs, splitting into chunks if necessary
func (g *GrpcFlagLogger) Write(ctx context.Context, request *resolverv1.WriteFlagLogsRequest) error {
	flagAssignedCount := len(request.FlagAssigned)
	clientResolveCount := len(request.ClientResolveInfo)
	flagResolveCount := len(request.FlagResolveInfo)

	if clientResolveCount == 0 && flagAssignedCount == 0 && flagResolveCount == 0 {
		g.logger.Debug("Skipping empty flag log request")
		return nil
	}

	if request.TelemetryData != nil {
		sdkID := "nil"
		sdkVersion := "nil"
		if request.TelemetryData.Sdk != nil {
			sdkID = request.TelemetryData.Sdk.GetId().String()
			sdkVersion = request.TelemetryData.Sdk.Version
		}
		g.logger.Info("Telemetry Data",
			"sdk_id", sdkID,
			"sdk_version", sdkVersion)
	}

	g.logger.Debug("Writing flag logs",
		"flag_assigned", flagAssignedCount,
		"client_resolve_info", clientResolveCount,
		"flag_resolve_info", flagResolveCount)

	// If flag_assigned list is small enough, send everything as-is
	if flagAssignedCount <= MaxFlagAssignedPerChunk {
		return g.sendAsync(ctx, request)
	}

	// Split flag_assigned into chunks and send each chunk asynchronously
	g.logger.Debug("Splitting flag_assigned entries into chunks",
		"total_entries", flagAssignedCount,
		"chunk_size", MaxFlagAssignedPerChunk)

	chunks := g.createFlagAssignedChunks(request)
	for _, chunk := range chunks {
		if err := g.sendAsync(ctx, chunk); err != nil {
			g.logger.Error("Failed to send flag log chunk", "error", err)
			return err
		}
	}

	return nil
}

// createFlagAssignedChunks splits the WriteFlagLogsRequest into chunks
func (g *GrpcFlagLogger) createFlagAssignedChunks(request *resolverv1.WriteFlagLogsRequest) []*resolverv1.WriteFlagLogsRequest {
	chunks := make([]*resolverv1.WriteFlagLogsRequest, 0)
	totalFlags := len(request.FlagAssigned)

	for i := 0; i < totalFlags; i += MaxFlagAssignedPerChunk {
		end := i + MaxFlagAssignedPerChunk
		if end > totalFlags {
			end = totalFlags
		}

		chunkBuilder := &resolverv1.WriteFlagLogsRequest{
			FlagAssigned: request.FlagAssigned[i:end],
		}

		// Include telemetry and resolve info only in the first chunk
		if i == 0 {
			if request.TelemetryData != nil {
				chunkBuilder.TelemetryData = request.TelemetryData
			}
			chunkBuilder.ClientResolveInfo = request.ClientResolveInfo
			chunkBuilder.FlagResolveInfo = request.FlagResolveInfo
		}

		chunks = append(chunks, chunkBuilder)
	}

	return chunks
}

func (g *GrpcFlagLogger) sendAsync(ctx context.Context, request *resolverv1.WriteFlagLogsRequest) error {
	g.wg.Add(1)
	go func() {
		defer g.wg.Done()
		// Create a context with timeout for the RPC
		rpcCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Add Authorization header with client secret
		md := metadata.Pairs("authorization", fmt.Sprintf("ClientSecret %s", g.clientSecret))
		rpcCtx = metadata.NewOutgoingContext(rpcCtx, md)

		if _, err := g.stub.ClientWriteFlagLogs(rpcCtx, request); err != nil {
			g.logger.Error("Failed to write flag logs", "error", err)
		} else {
			g.logger.Info("Successfully sent flag log", "entries", len(request.FlagAssigned))
		}
	}()
	return nil
}

// Shutdown waits for all pending async writes to complete
func (g *GrpcFlagLogger) Shutdown() {
	g.wg.Wait()
}

// NoOpWasmFlagLogger is a flag logger that drops all requests (for disabled logging)
type NoOpWasmFlagLogger struct{}

// Compile-time interface conformance check
var _ FlagLogger = (*NoOpWasmFlagLogger)(nil)

func NewNoOpWasmFlagLogger() *NoOpWasmFlagLogger {
	return &NoOpWasmFlagLogger{}
}

func (n *NoOpWasmFlagLogger) Write(ctx context.Context, request *resolverv1.WriteFlagLogsRequest) error {
	// Drop the request - do nothing
	return nil
}

func (n *NoOpWasmFlagLogger) Shutdown() {
	// Nothing to shut down
}

File: openfeature-provider/java/src/main/java/com/spotify/confidence/GrpcWasmFlagLogger.java
package com.spotify.confidence;

import static com.spotify.confidence.GrpcUtil.createConfidenceChannel;

import com.google.common.annotations.VisibleForTesting;
import com.spotify.confidence.flags.resolver.v1.InternalFlagLoggerServiceGrpc;
import com.spotify.confidence.flags.resolver.v1.WriteFlagLogsRequest;
import io.grpc.*;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@FunctionalInterface
interface FlagLogWriter {
  void write(WriteFlagLogsRequest request);
}

public class GrpcWasmFlagLogger implements WasmFlagLogger {
  private static final Logger logger = LoggerFactory.getLogger(GrpcWasmFlagLogger.class);
  // Max number of flag_assigned entries per chunk to avoid exceeding gRPC max message size
  private static final int MAX_FLAG_ASSIGNED_PER_CHUNK = 1000;
  private static final Duration DEFAULT_SHUTDOWN_TIMEOUT = Duration.ofSeconds(10);
  private final InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub stub;
  private final ExecutorService executorService;
  private final FlagLogWriter writer;
  private final Duration shutdownTimeout;
  private ManagedChannel channel;

  @VisibleForTesting
  public GrpcWasmFlagLogger(String clientSecret, FlagLogWriter writer) {
    this.stub = createAuthStub(new DefaultChannelFactory(), clientSecret);
    this.executorService = Executors.newCachedThreadPool();
    this.writer = writer;
    this.shutdownTimeout = DEFAULT_SHUTDOWN_TIMEOUT;
  }

  @VisibleForTesting
  public GrpcWasmFlagLogger(String clientSecret, FlagLogWriter writer, Duration shutdownTimeout) {
    this.stub = createAuthStub(new DefaultChannelFactory(), clientSecret);
    this.executorService = Executors.newCachedThreadPool();
    this.writer = writer;
    this.shutdownTimeout = shutdownTimeout;
  }

  public GrpcWasmFlagLogger(String clientSecret, ChannelFactory channelFactory) {
    this.stub = createAuthStub(channelFactory, clientSecret);
    this.executorService = Executors.newCachedThreadPool();
    this.shutdownTimeout = DEFAULT_SHUTDOWN_TIMEOUT;
    this.writer =
        request ->
            executorService.submit(
                () -> {
                  try {
                    stub.clientWriteFlagLogs(request);
                    logger.debug(
                        "Successfully sent flag log with {} entries",
                        request.getFlagAssignedCount());
                  } catch (Exception e) {
                    logger.error("Failed to write flag logs", e);
                  }
                });
  }

  private InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub createAuthStub(
      ChannelFactory channelFactory, String clientSecret) {
    this.channel = createConfidenceChannel(channelFactory);
    return addAuthInterceptor(InternalFlagLoggerServiceGrpc.newBlockingStub(channel), clientSecret);
  }

  @Override
  public void write(WriteFlagLogsRequest request) {
    if (request.getClientResolveInfoList().isEmpty()
        && request.getFlagAssignedList().isEmpty()
        && request.getFlagResolveInfoList().isEmpty()) {
      logger.debug("Skipping empty flag log request");
      return;
    }

    final int flagAssignedCount = request.getFlagAssignedCount();

    // If flag_assigned list is small enough, send everything as-is
    if (flagAssignedCount <= MAX_FLAG_ASSIGNED_PER_CHUNK) {
      sendAsync(request);
      return;
    }

    // Split flag_assigned into chunks and send each chunk asynchronously
    logger.debug(
        "Splitting {} flag_assigned entries into chunks of {}",
        flagAssignedCount,
        MAX_FLAG_ASSIGNED_PER_CHUNK);

    final List<WriteFlagLogsRequest> chunks = createFlagAssignedChunks(request);
    for (WriteFlagLogsRequest chunk : chunks) {
      sendAsync(chunk);
    }
  }

  private List<WriteFlagLogsRequest> createFlagAssignedChunks(WriteFlagLogsRequest request) {
    final List<WriteFlagLogsRequest> chunks = new ArrayList<>();
    final int totalFlags = request.getFlagAssignedCount();

    for (int i = 0; i < totalFlags; i += MAX_FLAG_ASSIGNED_PER_CHUNK) {
      final int end = Math.min(i + MAX_FLAG_ASSIGNED_PER_CHUNK, totalFlags);
      final WriteFlagLogsRequest.Builder chunkBuilder =
          WriteFlagLogsRequest.newBuilder()
              .addAllFlagAssigned(request.getFlagAssignedList().subList(i, end));

      // Include telemetry and resolve info only in the first chunk
      if (i == 0) {
        if (request.hasTelemetryData()) {
          chunkBuilder.setTelemetryData(request.getTelemetryData());
        }
        chunkBuilder
            .addAllClientResolveInfo(request.getClientResolveInfoList())
            .addAllFlagResolveInfo(request.getFlagResolveInfoList());
      }

      chunks.add(chunkBuilder.build());
    }

    return chunks;
  }

  private void sendAsync(WriteFlagLogsRequest request) {
    writer.write(request);
  }

  @Override
  public void writeSync(WriteFlagLogsRequest request) {
    if (request.getClientResolveInfoList().isEmpty()
        && request.getFlagAssignedList().isEmpty()
        && request.getFlagResolveInfoList().isEmpty()) {
      logger.debug("Skipping empty flag log request");
      return;
    }

    final int flagAssignedCount = request.getFlagAssignedCount();

    // If flag_assigned list is small enough, send everything as-is
    if (flagAssignedCount <= MAX_FLAG_ASSIGNED_PER_CHUNK) {
      sendSync(request);
      return;
    }

    // Split flag_assigned into chunks and send each chunk synchronously
    logger.debug(
        "Synchronously splitting {} flag_assigned entries into chunks of {}",
        flagAssignedCount,
        MAX_FLAG_ASSIGNED_PER_CHUNK);

    final List<WriteFlagLogsRequest> chunks = createFlagAssignedChunks(request);
    for (WriteFlagLogsRequest chunk : chunks) {
      sendSync(chunk);
    }
  }

  private void sendSync(WriteFlagLogsRequest request) {
    try {
      stub.clientWriteFlagLogs(request);
      logger.debug("Synchronously sent flag log with {} entries", request.getFlagAssignedCount());
    } catch (Exception e) {
      logger.error("Failed to write flag logs synchronously", e);
    }
  }

  /**
   * Shutdown the executor service and wait for pending async writes to complete. This method will
   * block for up to the configured shutdown timeout (default 10 seconds) waiting for pending log
   * writes to complete. Call this when the application is shutting down.
   */
  @Override
  public void shutdown() {
    if (channel != null) {
      channel.shutdown();
    }
    executorService.shutdown();
    try {
      if (!executorService.awaitTermination(shutdownTimeout.toMillis(), TimeUnit.MILLISECONDS)) {
        logger.warn(
            "Flag logger executor did not terminate within {} seconds, some logs may be lost",
            shutdownTimeout.getSeconds());
        executorService.shutdownNow();
      } else {
        logger.debug("Flag logger executor terminated gracefully");
      }
    } catch (InterruptedException e) {
      logger.warn("Interrupted while waiting for flag logger shutdown", e);
      executorService.shutdownNow();
      Thread.currentThread().interrupt();
    }
  }

  private static InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub
      addAuthInterceptor(
          InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub stub,
          String clientSecret) {
    // Create a stub with authorization header interceptor
    return stub.withInterceptors(
        new ClientInterceptor() {
          @Override
          public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
              MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
            return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
                next.newCall(method, callOptions)) {
              @Override
              public void start(Listener<RespT> responseListener, Metadata headers) {
                Metadata.Key<String> authKey =
                    Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
                headers.put(authKey, "ClientSecret " + clientSecret);
                super.start(responseListener, headers);
              }
            };
          }
        });
  }
}
